fixed-hash/src/hash.rs

```diff
@@ -6,735 +6,646 @@
+			pub fn zero() -> $name {
+				$name::repeat_byte(0u8)
 			}
-		}
 
-		impl $from {
-			/// Create a new, zero-initialised, instance.
-			#[deprecated(
-				since = "0.2.5",
-				note = "use `fixed_hash`::zero constructor instead"
-			)]
-			pub fn new() -> $from {
-				$from([0; $size])
+			/// Returns the size of this hash in bytes.
+			#[inline]
+			pub fn len_bytes() -> usize {
+				$n_bytes
 			}
 
-			/// Synonym for `new()`. Prefer to new as it's more readable.
-			pub fn zero() -> $from {
-				$from([0; $size])
+			/// Extracts a byte slice containing the entire fixed hash.
+			#[inline]
+			pub fn as_bytes(&self) -> &[u8] {
+				&self.0
 			}
 
-			/// Get the size of this object in bytes.
-			#[deprecated(
-				since = "0.2.5",
-				note = "will be renamed to `len_bytes` to avoid confusion"
-			)]
-			pub fn len() -> usize {
-				$size
+			/// Extracts a mutable byte slice containing the entire fixed hash.
+			#[inline]
+			pub fn as_bytes_mut(&mut self) -> &mut [u8] {
+				&mut self.0
 			}
 
-            /// Returns a constant raw pointer to the value
-            pub fn as_ptr(&self) -> *const u8 {
-                self.0.as_ptr()
-            }
-
-            pub fn as_mut_ptr(&mut self) -> *mut u8 {
-                (&mut self.0).as_mut_ptr()
-            }
-
+			/// Returns a constant raw pointer to the value.
 			#[inline]
-			/// Assign self to be of the same value as a slice of bytes of length `len()`.
-			#[deprecated(
-				since = "0.2.5",
-				note = "unconventional API, replaced by `assign_from_slice` in version 0.3"
-			)]
-			pub fn clone_from_slice(&mut self, src: &[u8]) -> usize {
-				let min = ::core::cmp::min($size, src.len());
-				self.0[..min].copy_from_slice(&src[..min]);
-				min
-			}
-
-			/// Convert a slice of bytes of length `len()` to an instance of this type.
-			#[deprecated(
-				since = "0.2.5",
-				note = "unconventional API, replaced by `new_from_slice` in version 0.3"
-			)]
-			pub fn from_slice(src: &[u8]) -> Self {
-				let mut r = Self::new();
-				r.clone_from_slice(src);
-				r
+			pub fn as_ptr(&self) -> *const u8 {
+				self.as_bytes().as_ptr()
 			}
 
-			/// Copy the data of this object into some mutable slice of length `len()`.
-			#[deprecated(
-				since = "0.2.5",
-				note = "use `std::slice` API instead"
-			)]
-			pub fn copy_to(&self, dest: &mut[u8]) {
-				let min = ::core::cmp::min($size, dest.len());
-				dest[..min].copy_from_slice(&self.0[..min]);
+			/// Returns a mutable raw pointer to the value.
+			#[inline]
+			pub fn as_mut_ptr(&mut self) -> *mut u8 {
+				self.as_bytes_mut().as_mut_ptr()
+			}
+
+			/// Assign the bytes from the byte slice `src` to `self`.
+			///
+			/// # Note
+			///
+			/// The given bytes are interpreted in big endian order.
+			///
+			/// # Panics
+			///
+			/// If the length of `src` and the number of bytes in `self` do not match.
+			pub fn assign_from_slice(&mut self, src: &[u8]) {
+				$crate::core::assert_eq!(src.len(), $n_bytes);
```

Comment:
I disagree very strongly with this argumentation to be honest. It was already a bad idea to disable assertions in C/C++ context at that time. What we could do is provide some `unsafe` APIs that very explicitely document this behaviour and only assert things in debug builds so that they stay debuggable.